### DISCUSSION TITLE
problems getting mcp23017 to work with IoAbstraction with interrupt (polling works though..)

### DISCUSSION THREAD

**Discussion by Hans0lo** [davetcc/IoAbstraction#183] [Q&A] [answered]
Hi,
for a ESP32 project I am working on I use a 4x4 button matrix.
I basically used the code found here:
https://www.thecoderscorner.com/jforum/posts/list/56.page

and that works just fine - but any attempt to get this working with an interrupt fails.

my HW setup in polling mode is the 8 pins from the button matrix connected to pins 1-8 of port A.

i then also connect the IntA output of my MCP23017 (adafruit breakout board) to ESP PIN 43 and made the following changes to the code:

```
IoAbstractionRef ioExpander = ioFrom23017(0x20);
//changed to:
IoAbstractionRef ioExpander = ioFrom23017(0x20,ACTIVE_LOW_OPEN,43);
```

```
keyboard.initialise(ioExpander, &keyLayout, &myListener);  //
//changed to
  keyboard.initialise(ioExpander, &keyLayout, &myListener, true);
```

sadly this did not work.

I tried it also with adding an isr function and these two lines:
```
pinMode(43, INPUT);
attachInterrupt(43,isr,CHANGE);
```

but this also did not work.

again, everything works in polling mode, but i would really love to get this working in interrupt mode

thnx in advance for any potential help!

**Comment by Hans0lo** [comment, marked as answer]
i dunno exactly why, as in the end everything now works perfectly... no ISR routine by me needs to be added, its all handled well by the lib.

IoAbstractionRef ioExpander = ioFrom23017(0x20,ACTIVE_LOW_OPEN,43);
keyboard.initialise(ioExpander, &keyLayout, &myListener, true);

since i have not really changed my initial approach I assume that maybe the interrupt cable was connected to a wrong pin on the ESP - that is at least the only idea i have why i did not work first, but does now.

thnx a million for this great lib!

**Reply by davetcc** [reply]
Apologies for not responding, for some reason I didn't see the notification due to a bit of noise in the inbox at the same time. But yes, that's right, there is no need for you to do anything as long as task managers run loop is called.

**Reply by Hans0lo** [reply]
now worries Dave, 
1. it worked out anyway, 
2. I learned some stuff on the way   :)

while it is sometimes faster and easier to get a question answered by a more experienced person, often I feel like I learn more if I have to dig deeper on my own compared to getting hand-holding  (also solving an issue then feels more like an accomplishment) 

I think a healthy combo of both is a good approach.

again, thnx for your work & support!
